libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp

```diff
@@ -0,0 +1,277 @@
+
+}
+
+
+uint8_t AP_InertialSensor_LSM6DS3::_register_read(uint8_t reg)
+{
+    uint8_t val = 0;
+	if (_sem->take_nonblocking()) 
+	{
+    	_dev->read_registers(reg, &val, 1);
+    }
+	return val;
+}
+
+void AP_InertialSensor_LSM6DS3::_register_write(uint8_t reg, uint8_t val, bool checked)
+{
+	if (_sem->take_nonblocking()) 
+	{
+    	_dev->write_register(reg, val);
+	}
+}
+
+void AP_InertialSensor_LSM6DS3::_gyro_init()
+{
+    _register_write(CTRL10_C,CTRL10_C_AXIS_VALUES);
+    _register_write(CTRL2_G,0x8C);
+    _set_gyro_scale(G_SCALE_2000DPS);
+}
+
+void AP_InertialSensor_LSM6DS3::_accel_init()
+{
+    _register_write(CTRL9_XL,CTRL9_XL_AXIS_VALUES);
+    hal.scheduler->delay(1);
+    _register_write(CTRL1_XL,CTRL1_XL_1600khz | CTRL1_XL_16g);
+    _set_accel_scale(A_SCALE_16G);
+    hal.scheduler->delay(1);
+}
+
+void AP_InertialSensor_LSM6DS3::_set_gyro_scale(gyro_scale scale)
+{
+    /* scales values from datasheet in mdps/digit */
+    switch (scale)
+	{
+	case G_SCALE_125DPS:
+        _gyro_scale = 4.375;
+        break;
+    case G_SCALE_245DPS:
+        _gyro_scale = 8.75;
+        break;
+    case G_SCALE_500DPS:
+        _gyro_scale = 17.50;
+        break;
+	case G_SCALE_1000DPS:
+        _gyro_scale = 35;
+        break;
+    case G_SCALE_2000DPS:
+        _gyro_scale = 70;
+        break;
+    }
+    // convert mdps/digit to dps/digit
+    _gyro_scale /= 1000;
+    //convert dps/digit to (rad/s)/digit
+    _gyro_scale *= DEG_TO_RAD;
+}
+
+void AP_InertialSensor_LSM6DS3::_set_accel_scale(accel_scale scale)
+{
+    /*
+     * Possible accelerometer scales (and their register bit settings) are:
+     * 2 g (00), 4g (10),  8g (11), 16g (01). Here's a bit of an
+     * algorithm to calculate g/(ADC tick) based on that 3-bit value:
+     */
+    _accel_scale = ((float) scale  / 32768.0f);
+
+    // convert to G/LSB to (m/s/s)/LSB */
+    _accel_scale *= GRAVITY_MSS;
+}
+
+/**
+ * Timer process to poll for new data from the LSM6DS3.
+ */
+void AP_InertialSensor_LSM6DS3::_poll_data()
+{
+    if (_gyro_data_ready())
+	{
+        _read_data_transaction_g();
+    }
+    if (_accel_data_ready())
+	{
+        _read_data_transaction_a();
+    }
+
+    // check next register value for correctness
+    if (!_dev->check_next_register())
+	{
+        _inc_gyro_error_count(_gyro_instance);
+    }
+    if (!_dev->check_next_register())
+	{
+        _inc_accel_error_count(_accel_instance);
+        
+    }
+}
+
+bool AP_InertialSensor_LSM6DS3::_accel_data_ready()
+{
+    uint8_t status = _register_read(STATUS_REG);
+    return status & XL_DATA_READY_MASK;
+}
+
+bool AP_InertialSensor_LSM6DS3::_gyro_data_ready()
+{
+    uint8_t status = _register_read(STATUS_REG);
+    return status & G_DATA_READY_MASK;
+}
+
+void AP_InertialSensor_LSM6DS3::_read_data_transaction_a()
+{
+    struct sensor_raw_data raw_data = {};
+    const uint8_t reg = OUT_X_L_A;
+    if (!_dev->transfer(&reg, 1, (uint8_t *) &raw_data, sizeof(raw_data)))
```

Comment:
can you use the sensors FIFO instead? That is generally a lot more efficient and avoids some nasty timing issues

Reply:
To use the sensors FIFO we should enable it in the FIFO_CTRL5 register, I don't completely follow what should be changed in those lines.